pyccel/codegen/printing/ccudacode.py

```diff
@@ -521,19 +552,18 @@ def copy_CudaArray_Data(self, expr):
         declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
         dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
         arg = rhs.arg if isinstance(rhs, (CudaArray, CupyArray)) else rhs
-        if rhs.rank > 1:
-            # flattening the args to use them in C initialization.
-            arg = self._flatten_list(arg)
 
-        self.add_import(c_imports['string'])
         if isinstance(arg, Variable):
             arg = self._print(arg)
-            cpy_data = "cudaMemcpy({0}.raw_data, {1}.{2}, {0}.buffer_size, cudaMemcpyHostToDevice);".format(lhs, arg, dtype)
+            cpy_data = f"cudaMemcpy({lhs}.{dtype}, {arg}.{dtype}, {lhs}.buffer_size, cudaMemcpyHostToDevice);"
             return '%s\n' % (cpy_data)
         else :
+            if arg.rank > 1:
+                arg = self._flatten_list(arg)
             arg = ', '.join(self._print(i) for i in arg)
             dummy_array = "%s %s[] = {%s};\n" % (declare_dtype, dummy_array_name, arg)
-            cpy_data = "cudaMemcpy({0}.raw_data, {1}, {0}.buffer_size, cudaMemcpyHostToDevice);".format(self._print(lhs), dummy_array_name, dtype)
+            target_array_name = self._print(lhs)
+            cpy_data = f"cudaMemcpy({target_array_name}.{dtype}, {dummy_array_name}, {target_array_name}.buffer_size, cudaMemcpyHostToDevice);"
```

Comment:
Beware, this way of writing the code does not allow for optional arrays

Reply:
I'm not really sure to understand what you mean by "optional arrays".

Reply:
See the discussion on this PR: https://github.com/pyccel/pyccel/pull/1280#discussion_r1058301708

Reply:
I'm sorry i took this late to answer.
if i understand correctly, this wont work if i had a pointer instead of a regular array ?
this is the case where the argument is not a variable and so it can't be a pointer, but i tried for the case of a variable.
Like the following case:
```python
@types('int[:]')
def func(arr = None):
    if arr is None:
        arr2 = cp.array([0, 1, 2, 3, 4])
    else:
        arr2 = cp.ravel(arr)
    return arr2 
```
I got the following result:
```C
extern "C" void func(t_ndarray *arr)
{
    t_ndarray arr2 = {.shape = NULL};
    if (arr == NULL)
    {

        int64_t tmp_shape[] = {5};
        arr2 = cuda_array_create(1, tmp_shape, nd_int64, false, allocateMemoryOnDevice);
        int64_t cuda_array_dummy[] = {0, 1, 2, 3, 4};
        cudaMemcpy(arr2.nd_int64, cuda_array_dummy, arr2.buffer_size, cudaMemcpyHostToDevice);
    }
    else
    {

        int64_t tmp_shape_0001[] = {arr->shape[0]};
        arr2 = cuda_array_create(1, tmp_shape_0001, nd_int64, false, allocateMemoryOnDevice);
        cudaMemcpy(arr2.nd_int64, (*arr).nd_int64, arr2.buffer_size, cudaMemcpyHostToDevice);
    }
}
```
I don't think there's a problem with this transpiration.

Reply:
> this is the case where the argument is not a variable and so it can't be a pointer, but i tried for the case of a variable.

That's why I just put "Beware" and didn't ask you to change it. I think it is better practice to always write it in a way that makes the pointer status implicit. That way you don't need to remember all the places where you need to use pointers and much less code needs updating if we add a case.

E.g. when code is added to be able to return cuda arrays, returning 2 arrays will make those arrays into pointers.

> I don't think there's a problem with this transpiration.

I see 2 problems:
1.  (Unrelated to this PR) The function says `return arr2` but `arr2` is not returned
2.  `(*arr).nd_int64` : this is an unidiomatic way of writing this. If class objects were used this would be written more idiomatically as `arr->nd_int64`. This is also a problem in the non-cuda code though. I need to open a PR and go through and sort it out. Probably I will do this when I write the codegen stage docs

Reply:
(By the way, transpiration is the liquid that your body generates when you're hot. I think you mean translation)

Reply:
I meant transpilation, i guess I clicked the auto correction without noticing 😅.